docs: design doc for LDAP intergration with RGW (#11320)
rook/rook: design/ceph/object/store.md

  zone:
    name: "name"
```

### LDAP

> The Ceph Object Gateway supports integrating with LDAP for authenticating and creating users, please refer [here](https://docs.ceph.com/docs/master/radosgw/ldap-auth/). This means that the `ceph object store user` is also required to be part of groups in the LDAP server otherwise, authentication will fail. The following settings need to be configured in the RGW server:

Reviewer: We probably should mention that CA for the x509 cert on ldaps is validated (AFAIK, there isn't a mechanism to disable validation) and a CA bundle will most likely need to be passed in to bind to ldaps.  E.g.

```yaml
kind: CephObjectStore
spec:
  gateway:
    caBundleRef: ldaps-cabundle
```

> ```

Reviewer: There is no mention of `rgw ldap searchfilter` in this block.  It isn't required to enable ldap support but I suspect it will be commonly used and a CRD should support it. Do we need to list the optional parameters in a separate subsection?

rgw ldap binddn =
rgw_ldap_secret = /etc/ceph/ldap/bindpass.secret
rgw ldap uri =
rgw ldap searchdn =
rgw ldap dnattr =
rgw_s3_auth_use_ldap = true
```
So the CRD for the Ceph Object Store will be modified to include the above changes:
```yaml

> ldap:

Reviewer 2: Are there any optional settings that the user may want to apply? If there are very many more I'd suggest we implement this as a `map[string]string` instead of defining them as individual string settings in the types.go, to allow for full flexibility of the ldap settings. But if this is the whole list, we don't need that flexibility.

Author: IMO the LDAP settings are minimal and straightforward, chances are less to have more options in future with RGW.
@jhoblitt @BlaineEXE any thoughts??

Reviewer: I believe `rgw_ldap_searchfilter` is the only practically optional parameter.  `rgw_ldap_secret` might technically be optional if it is possible to bind to the ldap server without auth.

  zone:
    name: "name"
```

### LDAP

> The Ceph Object Gateway supports integrating with LDAP for authenticating and creating users, please refer [here](https://docs.ceph.com/docs/master/radosgw/ldap-auth/). This means that the `ceph object store user` is also required to be part of groups in the LDAP server otherwise, authentication will fail. The following settings need to be configured in the RGW server:

Reviewer 3: > This means that the `ceph object store user` is also required to be part of groups in the LDAP server otherwise, authentication will fail.

Explain this more. What does this mean for CephObjectStoreUser resources and/or OBCs?

Reviewer: AFAIK, local users will still work as long as they do not conflict with ldap users.  I don't know which backend has priority if there is a conflict. I spent a few minutes scanning through and don't understand how `LocalEngine` and `LDAPEngine` hook in to the authentication flow. https://github.com/ceph/ceph/tree/02948ff7c61edca8b5e80e22806b69526370a066/src/rgw and don't see

```
rgw ldap binddn =
rgw_ldap_secret = /etc/ceph/ldap/bindpass.secret
rgw ldap uri =
rgw ldap searchdn =
rgw ldap dnattr =
rgw ldap searchfilter =

> rgw_s3_auth_use_ldap = true

Reviewer 3: Inconsistencies in underscores here. Why is that?

Reviewer: These are names used by `radosgw`: 

https://docs.ceph.com/en/quincy/radosgw/ldap-auth/#configuring-the-ceph-object-gateway-to-use-ldap-authentication

Reviewer: oh, you mean ` ` vs `_`. Yes, I agree. The names should match the literal config values.

```
So the CRD for the Ceph Object Store will be modified to include the above changes:
```yaml

> ldap:

Reviewer 3: Let's figure out where the ldap section should go in the CRD. for CephNFS, it is under the `security` heading. I'd propose this be `spec.security.ldap`

```suggestion
spec:
  security:
    ldap:
```

Reviewer: This sounds reasonable to me.

>   uri: ldaps://ldap-server:636
>   binddn: "uid=ceph,cn=users,cn=accounts,dc=example,dc=com"
>   credentialsecret: object-my-store-ldap-creds
>   searchdn: "cn=users,cn=accounts,dc=example,dc=com"
>   dnattr: "uid"
>   searchfilter: "memberof=cn=s3,cn=groups,cn=accounts,dc=example,dc=com"

Reviewer 3: Let's make sure there is a separation between configs that are direct 1:1 maps to RGW configs (suggested new `config` section) and configs that aren't (`credentialsecret`). For the credential secret, we will have the most ability to add different credential options in the future if we have a `credential` section.

```suggestion
  credential:
    secret:
      name: <k8s secret name>
  config:
    uri: ldaps://ldap-server:636
    binddn: "uid=ceph,cn=users,cn=accounts,dc=example,dc=com"
    searchdn: "cn=users,cn=accounts,dc=example,dc=com"
    dnattr: "uid"
    searchfilter: "memberof=cn=s3,cn=groups,cn=accounts,dc=example,dc=com"
```

We can easily support users who wish to inject credentials from vault if we allow users to omit the `credential` section. It also allows users to build the creds directly into the container image if they want. Here's how we documented that ability for NFS: https://github.com/BlaineEXE/rook/blob/f907852cac1096e51401247fb34467de79c87cee/Documentation/Storage-Configuration/NFS/nfs-security.md#L129-L132

Reviewer: I'm not very enthusiastic about naming config maps "config" but I could live with this.

```
* `uri`: It specifies the address of LDAP server to use.
* `binddn`: The bind domain for the service account used by RGW server.

> * `credentialsecret`: The name of k8s secret which contains password for the bind domain, need to save value as `data.password`.

Reviewer 3: ```suggestion
* `credentialSecret`: The name of k8s secret which contains password for the bind domain, need to save value as `data.password`.
```

  zone:
    name: "name"
```

### LDAP

The Ceph Object Gateway supports integrating with LDAP for authenticating and creating users, please refer [here](https://docs.ceph.com/docs/master/radosgw/ldap-auth/). This means that the `rgw backend user` is also required to be part of groups in the LDAP server otherwise, authentication will fail. The `rgw backend user` can be generated from `CephObjectStoreUser` or `ObjectBucketClaim` CRDs. For the both resources credentials are saved in Kubernetes Secrets which may not be valid with `LDAP Server`, user need to follow the steps mentioned [here](https://docs.ceph.com/en/latest/radosgw/ldap-auth/#generating-an-access-token-for-ldap-authentication).The following settings need to be configured in the RGW server:
```
rgw ldap binddn =
rgw ldap secret = /etc/ceph/ldap/bindpass.secret
rgw ldap uri =
rgw ldap searchdn =
rgw ldap dnattr =
rgw ldap searchfilter =

> rgw s3 auth use ldap = true

Reviewer: Why are the underscores being omitted here?  https://docs.ceph.com/en/latest/rados/configuration/ceph-conf/#option-names

Author: The options can be give with or without underscores in ceph.conf. Both works AFAIK

Reviewer: I didn't know that.

```
So the CRD for the Ceph Object Store will be modified to include the above changes:

> ```yaml

Reviewer 2: nit: blank lines before and after code blocks
```suggestion

```yaml
```

spec:
  security
    ldap:

>       config:

Reviewer 2: Could we just include the settings directly under `ldap` instead of adding another `config` element? Everything in the CRDs is conceptually config
```yaml
spec:
  security:
    ldap:
      uri: ldaps://ldap-server:636
      binddn: "uid=ceph,cn=users,cn=accounts,dc=example,dc=com"
      searchdn: "cn=users,cn=accounts,dc=example,dc=com"
      dnattr: "uid"
      searchfilter: "memberof=cn=s3,cn=groups,cn=accounts,dc=example,dc=com"
      credential:
        secret: object-my-store-ldap-creds
```

Reviewer 2: Or was there a need to define `config` as a generic `map[string]string` to allow for arbitrary settings to be passed to the config?

Reviewer: I agree with @travisn... `config` is a terrible name for a nested map.

Author: As per discussions made, sticking with the same options than the string map

        uri: ldaps://ldap-server:636
        binddn: "uid=ceph,cn=users,cn=accounts,dc=example,dc=com"
        searchdn: "cn=users,cn=accounts,dc=example,dc=com"
        dnattr: "uid"
        searchfilter: "memberof=cn=s3,cn=groups,cn=accounts,dc=example,dc=com"
      credential:

>         Secret: object-my-store-ldap-creds

Reviewer 2: ```suggestion
        secret: object-my-store-ldap-creds
```

```

> The `config` section includes options used for RGW wrt LDAP server

Reviewer 2: These will be strongly typed, right? We decided not to use a string map? If so, let's add that detail in the design doc.

* `uri`: It specifies the address of LDAP server to use.
* `binddn`: The bind domain for the service account used by RGW server.
* `searchdn`: The search domain where can it look for the user details.
* `dnattr`: The attribute being used in the constructed search filter to match a username, this can either be `uid` or `cn`.
* `searchfilter`: A generic search filter. If `dnattr` is set, this filter is `&()`'d together with the automatically constructed filter.

The `credential` section includes details related to authentication
* `secret`: The name of k8s secret which contains password for the bind domain, need to save value as `data.password`.
The CA bundle for ldap can be added to the `caBundleRef` option in `Gateway` settings:
```yaml

> 

Reviewer 2: nit: blank line before the yaml block instead of inside it

```
rgw ldap binddn =
rgw ldap secret = /etc/ceph/ldap/bindpass.secret
rgw ldap uri =
rgw ldap searchdn =
rgw ldap dnattr =
rgw ldap searchfilter =
rgw s3 auth use ldap = true
```
So the CRD for the Ceph Object Store will be modified to include the above changes:

```yaml
spec:
  security
    ldap:
      config:
        uri: ldaps://ldap-server:636
        binddn: "uid=ceph,cn=users,cn=accounts,dc=example,dc=com"
        searchdn: "cn=users,cn=accounts,dc=example,dc=com"
        dnattr: "uid"
        searchfilter: "memberof=cn=s3,cn=groups,cn=accounts,dc=example,dc=com"
      credential:
        secret: object-my-store-ldap-creds
```

> The `config` section includes options used for RGW wrt LDAP server. These options are strongly typed rather than string map approach since very less chance to modify in future.* `uri`: It specifies the address of LDAP server to use.

Reviewer 2: Need a line break
```suggestion
The `config` section includes options used for RGW wrt LDAP server. These options are strongly typed rather than string map approach since very less chance to modify in future.

* `uri`: It specifies the address of LDAP server to use.
```

```
rgw ldap binddn =
rgw ldap secret = /etc/ceph/ldap/bindpass.secret
rgw ldap uri =
rgw ldap searchdn =
rgw ldap dnattr =
rgw ldap searchfilter =
rgw s3 auth use ldap = true
```
So the CRD for the Ceph Object Store will be modified to include the above changes:

```yaml
spec:
  security
    ldap:
      config:
        uri: ldaps://ldap-server:636
        binddn: "uid=ceph,cn=users,cn=accounts,dc=example,dc=com"
        searchdn: "cn=users,cn=accounts,dc=example,dc=com"
        dnattr: "uid"
        searchfilter: "memberof=cn=s3,cn=groups,cn=accounts,dc=example,dc=com"
      credential:

>         secret: object-my-store-ldap-creds

Reviewer 3: For `credential`, let's design this to use the same mounting scheme that is working well for us in the CephNFS design that allows users to set the credential secret from whatever source is best for them.

See the `kerberos` section here (keytab is most similar to a cert):
https://github.com/rook/rook/blob/master/design/ceph/ceph-nfs-ganesha.md#example

See how this is documented here: https://rook.io/docs/rook/v1.11/CRDs/ceph-nfs-crd/#security
And here: https://rook.io/docs/rook/v1.11/Storage-Configuration/NFS/nfs-security/#kerberos-configuration

Users may omit the credential if they wish to inject it via vault or built it into their container statically. They may mount it from a K8s Secret, or from a directory on the host if that is where the LDAP creds are stored, as well as any other volume. 

This approach is highly flexible and means that we don't need to do extra work to add support for KMSes at a later date.

```

> The `config` section includes options used for RGW wrt LDAP server.Theses options are strongly typed than string map approach since very less chance to modify in future.

Reviewer 2: ```suggestion
The `config` section includes options used for RGW wrt LDAP server. These options are strongly typed rather than string map approach since very less chance to modify in future.
```

* `uri`: It specifies the address of LDAP server to use.* `binddn`: The bind domain for the service account used by RGW server.
* `searchdn`: The search domain where can it look for the user details.
* `dnattr`: The attribute being used in the constructed search filter to match a username, this can either be `uid` or `cn`.
* `searchfilter`: A generic search filter. If `dnattr` is set, this filter is `&()`'d together with the automatically constructed filter.

The `credential` defines where the password for accessing ldap server should be sourced from

* `volumeSource`: this is a standard Kubernetes VolumeSource for the Kerberos keytab file like
      what is normally used to configure Volumes for a Pod. For example, a Secret or HostPath.
      There are two requirements for the source's content:
    1. The config file must be mountable via `subPath: password`. For example, in a Secret, the
        data item must be named `password`, or `items` must be defined to select the key and
        give it path `password`. A HostPath directory must have the `password` file.

>     2. The volume or config file must have mode 0600.

Reviewer 3: When documenting this upon implementation, be sure to note where users should place the file `password` if they wish to do so manually.

Author: @jhoblitt ^^
